### PR TITLE
fix(realtime): replace boolean flags with _ResponseLifecycle enum

### PR DESCRIPTION
**Pull request by felmonon**
## Summary

Replaces the 4 boolean flags (`_ongoing_response`, `_response_cancel_pending`, `_response_create_pending`, `_queued_response_create`) with a single `_ResponseLifecycle` enum and one orthogonal boolean:

```python
class _ResponseLifecycle(enum.Enum):
    IDLE = "idle"
    CREATE_SENT = "create_sent"
    ACTIVE = "active"
    CANCEL_SENT = "cancel_sent"
```

**Why**: 4 booleans = 16 combinations, most invalid. The enum makes invalid states unrepresentable.

**Note**: The old boolean fields (`_ongoing_response`, `_response_cancel_pending`, `_response_create_pending`) have been **completely removed** from the codebase. All state is now tracked via `self._response_state: _ResponseLifecycle`.

### State transitions

```
IDLE ──send response.create──► CREATE_SENT
CREATE_SENT ──response.created──► ACTIVE
ACTIVE ──send response.cancel──► CANCEL_SENT
CANCEL_SENT ──response.done──► IDLE
ACTIVE ──response.done──► IDLE
CREATE_SENT ──error(rejected)──► IDLE
CANCEL_SENT ──error(rejected)──► IDLE
```

### Additional edge case fixes
- Failed `response.create` send now drains queued follow-ups instead of orphaning them
- Errors without `event_id` no longer wedge the session forever
- Raw/out-of-band `response.created` no longer clears SDK create tracking

## Test plan
- [x] All 72 existing tests updated and passing
- [x] 4 new tests for edge cases (send failure drain, error without event_id, raw response.created)
- [x] `ruff check` passes
- [ ] CI: Python 3.10–3.14, lint, typecheck

### CODE REVIEW

File: src/agents/realtime/openai_realtime.py [hunk 1]
```diff
@@ -674,9 +691,9 @@ async def close(self) -> None:
             self._websocket_task = None
 
     async def _cancel_response(self) -> None:
-        if self._ongoing_response:
+        if self._ongoing_response and not self._response_cancel_pending:
             await self._send_raw_message(OpenAIResponseCancelEvent(type="response.cancel"))
-            self._ongoing_response = False
+            self._response_cancel_pending = True
```
[the remark this answers is not in the document]
Comment (felmonon) [reply src/agents/realtime/openai_realtime.py]: Fixed in `5044356`. `_cancel_response()` now clears `_ongoing_response` and raises `_response_cancel_pending` before awaiting `response.cancel`, then restores the local state if that websocket send fails. I also added regressions for both the pre-await gate and the failure rollback.

File: src/agents/realtime/openai_realtime.py [hunk 2]
```diff
@@ -819,6 +836,13 @@ async def _handle_ws_event(self, event: dict[str, Any]):
             await self._emit_event(RealtimeModelTurnStartedEvent())
         elif parsed.type == "response.done":
             self._ongoing_response = False
+            queued = self._queued_response_create
+            self._response_cancel_pending = False
+            self._queued_response_create = False
+            if queued:
+                await self._send_raw_message(
+                    OpenAIResponseCreateEvent(type="response.create")
+                )
```
[the remark this answers is not in the document]
Comment (felmonon) [reply src/agents/realtime/openai_realtime.py]: Fixed in `5044356`. I split the start-response lifecycle so queued follow-ups stay gated across both the cancel window and the `response.create` -> `response.created` window. The drain path now preserves/requeues state on send failure, and the new tests cover in-flight drain requeueing, failed drain sends, queued sends while a start is still pending, and clearing that gate on `response.created`.

**Comment by felmonon** [comment]
@codex review

**Comment by felmonon** [comment]
@codex review

**Comment by felmonon** [comment]
@codex review

**Comment by felmonon** [comment]
@codex review

**Comment by felmonon** [comment]
@codex review

**Comment by felmonon** [comment]
@codex review

**Comment by felmonon** [comment]
@codex review

**Comment by felmonon** [comment]
@codex review

**Comment by felmonon** [comment]
Hey @seratch — I've been going back and forth with the Codex bot for a while now and I think we've hit a loop. Every round it raises the same ~7 comments, but they're all pointing at method signature lines without reading the method body where the fix already exists.

For example it keeps saying "clear _ongoing_response when starting a cancel" — but line 734 literally does `self._ongoing_response = False`. Same story for the other comments.

I've addressed every real issue it found along the way (drain ordering, flag-before-await races, error attribution via event_id tracking, stale response.done guards, raw send_event compat). The code and tests are solid at this point.

Would really appreciate a human look when you get a chance — happy to make any changes you want.

**Comment by seratch** [comment]
It seems you've been mentioning codex but this repo runs auto reviews for each commit. So, you don't need to do so. If you see duplicated comments, the double triggers (auto & your manual) could be the cause of it. Also, if the bot repeatedly point similar issues out, I believe the solution is not robust enough for various patterns. I don't think the bot could ignore your fixes (I've never observed such), so checking your code again and thinking of alternatives (if necessary) would be worth trying.

**Comment by seratch** [comment]
Also, while I haven't checked the code and its behavior in depth yet, at a glance, your code has a few states and it may be unnecessarily complex. I don't have anything specific to suggest yet, but having a few flag states may not be the direction we'd like to pursue.

**Comment by felmonon** [comment]
Totally agree — the boolean flags were the wrong approach. I replaced all 4 of them with a single `_ResponseLifecycle` enum: `IDLE → CREATE_SENT → ACTIVE → CANCEL_SENT`. Invalid states are now structurally impossible instead of guarded by logic. The diff is actually 19 lines shorter.

Won't trigger Codex manually anymore either, sorry about the noise.

**Comment by felmonon** [comment]
Hey @seratch, quick update on where things stand.

I've gone through about 8 rounds of fixes based on Codex's feedback. The big stuff is all handled  the enum state machine replaced the 4 booleans, stale response.done events get ignored via response ID tracking, cancel correlation works correctly, the session won't wedge on no-event-id errors, and close() cleans up in the right order.

Codex just posted 5 more comments. One P1 that's actually already fixed in the current code (it flagged the exact thing I added in the last commit). The other four are P2s about pretty exotic scenarios  things like what happens if you mix raw websocket sends with SDK-managed sends while the socket dies mid-send, or if the server rejects a retried create twice in a row. Some of these are protocol limitations too response.created doesn't echo the client event_id, so there's no way to perfectly correlate in mixed-send scenarios.

I think Codex is going to keep finding increasingly niche interleavings on state machine code like this. The real-world correctness issues from the original design are fixed. Want me to keep going on these edge cases or is this in good shape for review?

**Comment by seratch** [comment]
Thanks for sharing the update. Indeed, Codex can sometimes mention edge cases, but considering the latest code here, which introduces a few more internal states, I am not surprised to see Codex points more things out. I don't feel skipping the comments from Codex for this specific code change suggestion is a good idea.

Since I have other priorities right now, I don't have the bandwidth to check the suggestion in depth or come up with my own alternative quickly. If your projects are not directly affected by this issue, please hold off working on this PR.

**Comment by felmonon** [comment]
Totally fair I'll hold off. And yeah, good point about the added states making Codex noisier. If you end up wanting a simpler direction when you have the bandwidth, I'm happy to rework it. No rush on my end.

**Comment by felmonon** [comment]
Closing this draft in favor of #2754, which takes a smaller session-layer approach to the guardrail interrupt race and passes the full realtime test suite.
